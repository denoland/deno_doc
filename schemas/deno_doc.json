{
  "$id": "https://deno.land/schemas/deno_doc.json",
  "$schema": "http://json-schema.org/draft-07/schema",
  "description": "A JSON representation of a deno_doc documentation structure",
  "type": "array",
  "items": {
    "$ref": "#/definitions/DocNode"
  },
  "definitions": {
    "DocNode": {
      "type": "object",
      "required": [
        "kind",
        "name",
        "location"
      ],
      "properties": {
        "kind": {
          "type": "string",
          "enum": [
            "function",
            "variable",
            "enum",
            "class",
            "typeAlias",
            "namespace",
            "interface",
            "import"
          ]
        },
        "name": {
          "type": "string"
        },
        "location": {
          "$ref": "#/definitions/Location"
        },
        "jsDoc": {
          "type": "string"
        }
      },
      "anyOf": [
        {
          "required": [
            "functionDef"
          ],
          "properties": {
            "kind": {
              "const": "function"
            },
            "functionDef": {
              "$ref": "#/definitions/FunctionDef"
            }
          }
        },
        {
          "required": [
            "variableDef"
          ],
          "properties": {
            "kind": {
              "const": "variable"
            },
            "variableDef": {
              "$ref": "#/definitions/VariableDef"
            }
          }
        },
        {
          "required": [
            "enumDef"
          ],
          "properties": {
            "kind": {
              "const": "enum"
            },
            "enumDef": {
              "$ref": "#/definitions/EnumDef"
            }
          }
        },
        {
          "required": [
            "classDef"
          ],
          "properties": {
            "kind": {
              "const": "class"
            }
          },
          "classDef": {
            "$ref": "#/definitions/ClassDef"
          }
        },
        {
          "required": [
            "typeAliasDef"
          ],
          "properties": {
            "kind": {
              "const": "typeAlias"
            }
          },
          "typeAliasDef": {
            "$ref": "#/definitions/TypeAliasDef"
          }
        },
        {
          "required": [
            "namespaceDef"
          ],
          "properties": {
            "kind": {
              "const": "namespace"
            }
          },
          "namespaceDef": {
            "$ref": "#/definitions/NamespaceDef"
          }
        },
        {
          "required": [
            "interfaceDef"
          ],
          "properties": {
            "kind": {
              "const": "interface"
            }
          },
          "interfaceDef": {
            "$ref": "#/definitions/interfaceDef"
          }
        },
        {
          "required": [
            "importDef"
          ],
          "properties": {
            "kind": {
              "const": "import"
            },
            "importDef": {
              "$ref": "#/definitions/ImportDef"
            }
          }
        }
      ]
    },
    "Location": {
      "type": "object",
      "required": [
        "filename",
        "line",
        "col"
      ],
      "properties": {
        "filename": {
          "type": "string"
        },
        "line": {
          "type": "number"
        },
        "col": {
          "type": "number"
        }
      }
    }
  }
}
