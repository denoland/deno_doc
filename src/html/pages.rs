use super::sidepanels;
use super::sidepanels::SidepanelCtx;
use super::symbols::SymbolContentCtx;
use super::util::BreadcrumbsCtx;
use super::util::SectionHeaderCtx;
use super::DocNodeWithContext;
use super::FileMode;
use super::GenerateCtx;
use super::RenderContext;
use super::ShortPath;
use super::SymbolGroupCtx;
use super::UrlResolveKind;
use std::rc::Rc;

use super::FUSE_FILENAME;
use super::PAGE_STYLESHEET_FILENAME;
use super::SCRIPT_FILENAME;
use super::SEARCH_FILENAME;
use super::SEARCH_INDEX_FILENAME;
use super::STYLESHEET_FILENAME;

use crate::DocNode;
use crate::DocNodeKind;
use indexmap::IndexMap;
use serde::Serialize;

#[derive(Debug, Serialize, Clone)]
pub struct HtmlHeadCtx {
  title: String,
  current_file: String,
  stylesheet_url: String,
  page_stylesheet_url: String,
  url_search_index: String,
  script_js: String,
  fuse_js: String,
  url_search: String,
}

impl HtmlHeadCtx {
  pub const TEMPLATE: &'static str = "pages/html_head";

  pub fn new(
    root: &str,
    page: &str,
    package_name: Option<&String>,
    current_file: Option<&ShortPath>,
  ) -> Self {
    Self {
      title: format!(
        "{page} - {}documentation",
        package_name
          .map(|package_name| format!("{package_name} "))
          .unwrap_or_default()
      ),
      current_file: current_file
        .map(|current_file| &*current_file.path)
        .unwrap_or_default()
        .to_string(),
      stylesheet_url: format!("{root}{STYLESHEET_FILENAME}"),
      page_stylesheet_url: format!("{root}{PAGE_STYLESHEET_FILENAME}"),
      url_search_index: format!("{root}{SEARCH_INDEX_FILENAME}"),
      script_js: format!("{root}{SCRIPT_FILENAME}"),
      fuse_js: format!("{root}{FUSE_FILENAME}"),
      url_search: format!("{root}{SEARCH_FILENAME}"),
    }
  }
}

#[derive(Debug, Serialize, Clone)]
pub struct IndexCtx {
  pub html_head_ctx: HtmlHeadCtx,
  pub sidepanel_ctx: sidepanels::IndexSidepanelCtx,
  pub module_doc: Option<super::jsdoc::ModuleDocCtx>,
  pub all_symbols: Option<SymbolContentCtx>,
  pub breadcrumbs_ctx: BreadcrumbsCtx,
}

impl IndexCtx {
  pub const TEMPLATE: &'static str = "pages/index";

  /// short_path is None in the case this is a root index page but there is no main entrypoint
  pub fn new(
    ctx: &GenerateCtx,
    short_path: Option<Rc<ShortPath>>,
    partitions: super::partition::Partitions<String>,
  ) -> Self {
    // will be default on index page with no main entrypoint
    let default = vec![];
    let doc_nodes = short_path
      .as_ref()
      .and_then(|short_path| ctx.doc_nodes.get(short_path))
      .unwrap_or(&default);

    let render_ctx = RenderContext::new(
      ctx,
      doc_nodes,
      short_path
        .as_deref()
        .map_or(UrlResolveKind::Root, ShortPath::as_resolve_kind),
    );

    let module_doc = short_path.as_ref().map(|short_path| {
      super::jsdoc::ModuleDocCtx::new(&render_ctx, short_path)
    });

    let root = ctx.href_resolver.resolve_path(
      short_path
        .as_deref()
        .map_or(UrlResolveKind::Root, ShortPath::as_resolve_kind),
      UrlResolveKind::Root,
    );

    let html_head_ctx =
      HtmlHeadCtx::new(&root, "Index", ctx.package_name.as_ref(), None);

    let all_symbols = if ctx.file_mode == FileMode::SingleDts {
      let sections = super::namespace::render_namespace(
        &render_ctx,
        partitions
          .clone()
          .into_iter()
          .map(|(title, nodes)| {
            (
              SectionHeaderCtx {
                title,
                href: None,
                doc: None,
              },
              nodes,
            )
          })
          .collect(),
      );

      Some(SymbolContentCtx {
        id: String::new(),
        sections,
        docs: None,
      })
    } else {
      None
    };

    let sidepanel_ctx =
      sidepanels::IndexSidepanelCtx::new(ctx, short_path.clone(), partitions);

    IndexCtx {
      html_head_ctx,
      sidepanel_ctx,
      module_doc,
      all_symbols,
      breadcrumbs_ctx: render_ctx.get_breadcrumbs(),
    }
  }
}

#[derive(Serialize)]
pub struct AllSymbolsCtx {
  pub html_head_ctx: HtmlHeadCtx,
  pub content: SymbolContentCtx,
  pub breadcrumbs_ctx: BreadcrumbsCtx,
}

impl AllSymbolsCtx {
  pub const TEMPLATE: &'static str = "pages/all_symbols";

  pub fn new(
    ctx: &GenerateCtx,
    partitions: super::partition::Partitions<Rc<ShortPath>>,
  ) -> Self {
    // TODO(@crowlKats): handle doc_nodes in all symbols page for each symbol
    let render_ctx = RenderContext::new(ctx, &[], UrlResolveKind::AllSymbols);

    let sections = super::namespace::render_namespace(
      &render_ctx,
      partitions
        .into_iter()
        .map(|(path, nodes)| {
          (
            SectionHeaderCtx::new_for_namespace(&render_ctx, &path),
            nodes,
          )
        })
        .collect(),
    );

    let html_head_ctx =
      HtmlHeadCtx::new("./", "All Symbols", ctx.package_name.as_ref(), None);

    AllSymbolsCtx {
      html_head_ctx,
      content: SymbolContentCtx {
        id: String::new(),
        sections,
        docs: None,
      },
      breadcrumbs_ctx: render_ctx.get_breadcrumbs(),
    }
  }
}

pub enum SymbolPage {
  Symbol {
    breadcrumbs_ctx: BreadcrumbsCtx,
    sidepanel_ctx: SidepanelCtx,
    symbol_group_ctx: SymbolGroupCtx,
  },
  Redirect {
    current_symbol: String,
    href: String,
  },
}

pub fn generate_symbol_pages_for_module(
  ctx: &GenerateCtx,
  short_path: &ShortPath,
  sidepanel_partitions: &super::partition::Partitions<String>,
  module_doc_nodes: &[DocNodeWithContext],
) -> Vec<SymbolPage> {
  let mut name_partitions =
    super::partition::partition_nodes_by_name(module_doc_nodes, true);

  let mut drilldown_partitions = IndexMap::new();
  for doc_nodes in name_partitions.values() {
    let has_class =
      doc_nodes.iter().any(|node| node.kind == DocNodeKind::Class);
    for doc_node in doc_nodes {
      match doc_node.kind {
        DocNodeKind::Class => {
          let class = doc_node.class_def.as_ref().unwrap();

          let method_nodes = class
            .methods
            .iter()
            .map(|method| {
              doc_node.create_child_method(
                DocNode::function(
                  method.name.clone(),
                  method.location.clone(),
                  doc_node.declaration_kind,
                  method.js_doc.clone(),
                  method.function_def.clone(),
                ),
                method.is_static,
              )
            })
            .collect::<Vec<_>>();

          drilldown_partitions.extend(
            super::partition::partition_nodes_by_name(&method_nodes, false),
          );

          let property_nodes = class
            .properties
            .iter()
            .map(|property| {
              doc_node.create_child_property(
                DocNode::from(property.clone()),
                property.is_static,
              )
            })
            .collect::<Vec<_>>();

          drilldown_partitions.extend(
            super::partition::partition_nodes_by_name(&property_nodes, false),
          );
        }
        DocNodeKind::Interface => {
          let interface = doc_node.interface_def.as_ref().unwrap();
          let method_nodes = interface
            .methods
            .iter()
            .filter_map(|method| {
              if has_class && method.name == "prototype" {
                None
              } else {
                Some(
                  doc_node
                    .create_child_method(DocNode::from(method.clone()), true),
                )
              }
            })
            .collect::<Vec<_>>();

          drilldown_partitions.extend(
            super::partition::partition_nodes_by_name(&method_nodes, false),
          );

          let property_nodes =
            interface
              .properties
              .iter()
              .filter_map(|property| {
                if has_class && property.name == "prototype" {
                  None
                } else {
                  Some(doc_node.create_child_property(
                    DocNode::from(property.clone()),
                    true,
                  ))
                }
              })
              .collect::<Vec<_>>();

          drilldown_partitions.extend(
            super::partition::partition_nodes_by_name(&property_nodes, false),
          );
        }
        DocNodeKind::TypeAlias => {
          let type_alias = doc_node.type_alias_def.as_ref().unwrap();

          if let Some(ts_type_literal) =
            type_alias.ts_type.type_literal.as_ref()
          {
            let method_nodes = ts_type_literal
              .methods
              .iter()
              .filter_map(|method| {
                if has_class && method.name == "prototype" {
                  None
                } else {
                  Some(
                    doc_node
                      .create_child_method(DocNode::from(method.clone()), true),
                  )
                }
              })
              .collect::<Vec<_>>();

            drilldown_partitions.extend(
              super::partition::partition_nodes_by_name(&method_nodes, false),
            );

            let property_nodes = ts_type_literal
              .properties
              .iter()
              .filter_map(|property| {
                if has_class && property.name == "prototype" {
                  None
                } else {
                  Some(doc_node.create_child_property(
                    DocNode::from(property.clone()),
                    true,
                  ))
                }
              })
              .collect::<Vec<_>>();

            drilldown_partitions.extend(
              super::partition::partition_nodes_by_name(&property_nodes, false),
            );
          }
        }
        DocNodeKind::Variable => {
          let variable = doc_node.variable_def.as_ref().unwrap();

          if let Some(ts_type_literal) = variable
            .ts_type
            .as_ref()
            .and_then(|ts_type| ts_type.type_literal.as_ref())
          {
            let method_nodes = ts_type_literal
              .methods
              .iter()
              .map(|method| {
                doc_node
                  .create_child_method(DocNode::from(method.clone()), true)
              })
              .collect::<Vec<_>>();

            drilldown_partitions.extend(
              super::partition::partition_nodes_by_name(&method_nodes, false),
            );

            let property_nodes = ts_type_literal
              .properties
              .iter()
              .map(|property| {
                doc_node
                  .create_child_property(DocNode::from(property.clone()), true)
              })
              .collect::<Vec<_>>();

            drilldown_partitions.extend(
              super::partition::partition_nodes_by_name(&property_nodes, false),
            );
          }
        }
        _ => {}
      }
    }
  }
  name_partitions.extend(drilldown_partitions);

  let mut generated_pages = Vec::with_capacity(name_partitions.values().len());

  let render_ctx =
    RenderContext::new(ctx, module_doc_nodes, UrlResolveKind::File(short_path));

  for (name, doc_nodes) in name_partitions {
    let sidepanel_ctx =
      SidepanelCtx::new(ctx, sidepanel_partitions, short_path, &name);

    let (breadcrumbs_ctx, symbol_group_ctx) =
      render_symbol_page(&render_ctx, short_path, &name, &doc_nodes);

    generated_pages.push(SymbolPage::Symbol {
      breadcrumbs_ctx,
      sidepanel_ctx,
      symbol_group_ctx,
    });

    if doc_nodes
      .iter()
      .any(|doc_node| doc_node.kind == DocNodeKind::Class)
    {
      let prototype_name = format!("{name}.prototype");
      generated_pages.push(SymbolPage::Redirect {
        href: ctx.href_resolver.resolve_path(
          UrlResolveKind::Symbol {
            file: short_path,
            symbol: &prototype_name,
          },
          UrlResolveKind::Symbol {
            file: short_path,
            symbol: &name,
          },
        ),
        current_symbol: prototype_name,
      });
    }
  }

  generated_pages
}

#[derive(Debug, Serialize)]
pub struct SymbolPageCtx {
  pub html_head_ctx: HtmlHeadCtx,
  pub sidepanel_ctx: SidepanelCtx,
  pub symbol_group_ctx: SymbolGroupCtx,
  pub breadcrumbs_ctx: BreadcrumbsCtx,
}

impl SymbolPageCtx {
  pub const TEMPLATE: &'static str = "pages/symbol";
}

pub fn render_symbol_page(
  render_ctx: &RenderContext,
  short_path: &ShortPath,
  namespaced_name: &str,
  doc_nodes: &[DocNodeWithContext],
) -> (BreadcrumbsCtx, SymbolGroupCtx) {
  let mut render_ctx =
    render_ctx.with_current_resolve(UrlResolveKind::Symbol {
      file: short_path,
      symbol: namespaced_name,
    });
  if !doc_nodes[0].ns_qualifiers.is_empty() {
    render_ctx = render_ctx.with_namespace(doc_nodes[0].ns_qualifiers.clone());
  }

  // NOTE: `doc_nodes` should be sorted at this point.
  let symbol_group_ctx =
    SymbolGroupCtx::new(&render_ctx, doc_nodes, namespaced_name);

  (render_ctx.get_breadcrumbs(), symbol_group_ctx)
}
