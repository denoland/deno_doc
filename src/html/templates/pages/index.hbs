{{~> pages/html_head html_head_ctx ~}}
{{~> category_panel categories_panel ~}}
<div>
  {{~> pages/top_nav ~}}

  <div id="content">
    <main>
      {{~#if usage~}}
        {{~> usages_large usage ~}}
      {{~/if~}}

      {{~#if module_doc~}}
        {{~> module_doc module_doc ~}}
      {{~/if~}}

      {{~#if overview~}}
        {{~> symbol_content overview ~}}
      {{~/if~}}
    </main>

    {{~> toc toc_ctx ~}}

    {{~> pages/search_results ~}}
  </div>
</div>
</div>
</body>
</html>
