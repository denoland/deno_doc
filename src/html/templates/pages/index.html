{{ call pages/html_head.html with html_head_ctx }}

{{ call index_sidepanel.html with sidepanel_ctx }}

<div id="content" class="with_sidepanel">
  {{ call breadcrumbs.html with breadcrumbs_ctx }}
  {{ call pages/search_bar.html with search_ctx }}

  <main>
    {#- TODO(bartlomieju): call into template here #}

    {{ if module_doc }}
      {{ call module_doc.html with module_doc }}
    {{ endif }}
  </main>

  {{ call pages/search_results.html with search_ctx }}
<div>
</body>
</html>
