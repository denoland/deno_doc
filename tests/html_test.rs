// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.

use deno_ast::ModuleSpecifier;
use deno_doc::html::pages::SymbolPage;
use deno_doc::html::*;
use deno_doc::DocNode;
use deno_doc::DocParser;
use deno_doc::DocParserOptions;
use deno_graph::source::LoadFuture;
use deno_graph::source::LoadOptions;
use deno_graph::source::LoadResponse;
use deno_graph::source::Loader;
use deno_graph::BuildOptions;
use deno_graph::CapturingModuleAnalyzer;
use deno_graph::GraphKind;
use deno_graph::ModuleGraph;
use futures::future;
use indexmap::IndexMap;
use std::fs;
use std::rc::Rc;

struct SourceFileLoader {}

impl Loader for SourceFileLoader {
  fn load(
    &self,
    specifier: &ModuleSpecifier,
    _options: LoadOptions,
  ) -> LoadFuture {
    let result = if specifier.scheme() == "file" {
      let path = specifier.to_file_path().unwrap();
      fs::read(path)
        .map(|content| {
          Some(LoadResponse::Module {
            specifier: specifier.clone(),
            maybe_headers: None,
            content: content.into(),
          })
        })
        .map_err(|err| err.into())
    } else {
      Ok(None)
    };
    Box::pin(future::ready(result))
  }
}

struct EmptyResolver {}

impl HrefResolver for EmptyResolver {
  fn resolve_path(
    &self,
    current: UrlResolveKind,
    target: UrlResolveKind,
  ) -> String {
    href_path_resolve(current, target)
  }

  fn resolve_global_symbol(&self, _symbol: &[String]) -> Option<String> {
    None
  }

  fn resolve_import_href(
    &self,
    _symbol: &[String],
    _src: &str,
  ) -> Option<String> {
    None
  }

  fn resolve_usage(&self, current_resolve: UrlResolveKind) -> Option<String> {
    current_resolve
      .get_file()
      .map(|current_file| current_file.path.to_string())
  }

  fn resolve_source(&self, _location: &deno_doc::Location) -> Option<String> {
    None
  }
}

async fn get_files(subpath: &str) -> IndexMap<ModuleSpecifier, Vec<DocNode>> {
  let files = fs::read_dir(
    std::env::current_dir()
      .unwrap()
      .join("tests")
      .join("testdata")
      .join(subpath),
  )
  .unwrap();

  let source_files: Vec<ModuleSpecifier> = files
    .into_iter()
    .map(|entry| {
      let entry = entry.unwrap();
      ModuleSpecifier::from_file_path(entry.path()).unwrap()
    })
    .collect();

  let loader = SourceFileLoader {};
  let analyzer = CapturingModuleAnalyzer::default();
  let mut graph = ModuleGraph::new(GraphKind::TypesOnly);
  graph
    .build(
      source_files.clone(),
      &loader,
      BuildOptions {
        module_analyzer: &analyzer,
        ..Default::default()
      },
    )
    .await;

  let parser = DocParser::new(
    &graph,
    &analyzer,
    DocParserOptions {
      diagnostics: false,
      private: false,
    },
  )
  .unwrap();

  let mut source_files = source_files.clone();
  source_files.sort();
  let mut doc_nodes_by_url = IndexMap::with_capacity(source_files.len());
  for source_file in source_files {
    let nodes = parser.parse_with_reexports(&source_file).unwrap();
    doc_nodes_by_url.insert(source_file, nodes);
  }

  doc_nodes_by_url
}

#[tokio::test]
async fn html_doc_files() {
  let files = generate(
    GenerateOptions {
      package_name: None,
      main_entrypoint: None,
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: None,
      composable_output: false,
    },
    get_files("single").await,
  )
  .unwrap();

  let mut file_names = files.keys().collect::<Vec<_>>();
  file_names.sort();

  assert_eq!(
    file_names,
    [
      "./all_symbols.html",
      "./index.html",
      "./~/Bar.html",
      "./~/Bar.prototype.html",
      "./~/Foo.html",
      "./~/Foo.prototype.html",
      "./~/Foobar.html",
      "./~/Foobar.prototype.html",
      "fuse.js",
      "page.css",
      "script.js",
      "search.js",
      "search_index.js",
      "styles.css",
    ]
  );

  #[cfg(all(not(feature = "syntect"), not(feature = "tree-sitter")))]
  {
    insta::assert_snapshot!(files.get("./all_symbols.html").unwrap());
    insta::assert_snapshot!(files.get("./index.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Bar.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Bar.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foobar.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foobar.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("fuse.js").unwrap());
    insta::assert_snapshot!(files.get("page.css").unwrap());
    insta::assert_snapshot!(files.get("script.js").unwrap());
    insta::assert_snapshot!(files.get("search.js").unwrap());
    insta::assert_snapshot!(files.get("search_index.js").unwrap());
  }
}

#[tokio::test]
async fn html_doc_files_rewrite() {
  let multiple_dir = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("multiple");
  let mut rewrite_map = IndexMap::new();
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
    ".".to_string(),
  );
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("b.ts")).unwrap(),
    "foo".to_string(),
  );

  let files = generate(
    GenerateOptions {
      package_name: None,
      main_entrypoint: None,
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: Some(rewrite_map),
      composable_output: false,
    },
    get_files("multiple").await,
  )
  .unwrap();

  let mut file_names = files.keys().collect::<Vec<_>>();
  file_names.sort();

  assert_eq!(
    file_names,
    [
      "./all_symbols.html",
      "./index.html",
      "./~/A.html",
      "./~/A.prototype.html",
      "./~/B.html",
      "./~/B.prototype.html",
      "./~/Bar.html",
      "./~/Bar.prototype.html",
      "./~/Baz.foo.html",
      "./~/Baz.html",
      "./~/Foo.bar.html",
      "./~/Foo.html",
      "./~/Foo.prototype.\"><img src=x onerror=alert(1)>.html",
      "./~/Foo.prototype.foo.html",
      "./~/Foo.prototype.html",
      "./~/Foobar.html",
      "./~/Foobar.prototype.html",
      "./~/Hello.html",
      "./~/Hello.world.html",
      "./~/index.html",
      "foo/~/index.html",
      "foo/~/x.html",
      "fuse.js",
      "page.css",
      "script.js",
      "search.js",
      "search_index.js",
      "styles.css"
    ]
  );

  #[cfg(all(not(feature = "syntect"), not(feature = "tree-sitter")))]
  {
    insta::assert_snapshot!(files.get("./all_symbols.html").unwrap());
    insta::assert_snapshot!(files.get("./index.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Bar.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Bar.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Baz.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Baz.foo.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.bar.html").unwrap());
    insta::assert_snapshot!(files
      .get("./~/Foo.prototype.\"><img src=x onerror=alert(1)>.html")
      .unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.prototype.foo.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foo.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foobar.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Foobar.prototype.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Hello.html").unwrap());
    insta::assert_snapshot!(files.get("./~/Hello.world.html").unwrap());
    insta::assert_snapshot!(files.get("./~/index.html").unwrap());
    insta::assert_snapshot!(files.get("foo/~/index.html").unwrap());
    insta::assert_snapshot!(files.get("foo/~/x.html").unwrap());
    insta::assert_snapshot!(files.get("fuse.js").unwrap());
    insta::assert_snapshot!(files.get("page.css").unwrap());
    insta::assert_snapshot!(files.get("script.js").unwrap());
    insta::assert_snapshot!(files.get("search.js").unwrap());
    insta::assert_snapshot!(files.get("search_index.js").unwrap());
  }
}

#[tokio::test]
async fn symbol_group() {
  let multiple_dir = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("multiple");

  let doc_nodes_by_url = get_files("multiple").await;

  let mut rewrite_map = IndexMap::new();
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
    ".".to_string(),
  );
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("b.ts")).unwrap(),
    "foo".to_string(),
  );

  let ctx = GenerateCtx::new(
    GenerateOptions {
      package_name: None,
      main_entrypoint: Some(
        ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
      ),
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: Some(rewrite_map),
      composable_output: false,
    },
    None,
    Default::default(),
    doc_nodes_by_url,
  )
  .unwrap();

  let mut files = vec![];

  {
    for (short_path, doc_nodes) in &ctx.doc_nodes {
      let symbol_pages =
        generate_symbol_pages_for_module(&ctx, short_path, doc_nodes);

      files.extend(symbol_pages.into_iter().map(
        |symbol_page| match symbol_page {
          SymbolPage::Symbol {
            breadcrumbs_ctx,
            symbol_group_ctx,
            toc_ctx,
          } => {
            let root = ctx.href_resolver.resolve_path(
              UrlResolveKind::Symbol {
                file: short_path,
                symbol: &symbol_group_ctx.name,
              },
              UrlResolveKind::Root,
            );

            let html_head_ctx = pages::HtmlHeadCtx::new(
              &root,
              &symbol_group_ctx.name,
              ctx.package_name.as_ref(),
              Some(short_path),
            );

            Some(pages::SymbolPageCtx {
              html_head_ctx,
              symbol_group_ctx,
              breadcrumbs_ctx,
              toc_ctx,
            })
          }
          SymbolPage::Redirect { .. } => None,
        },
      ));
    }
  }

  let mut files_json = serde_json::to_string_pretty(&files).unwrap();
  files_json.push('\n');

  let testdata_path = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata");

  #[cfg(feature = "syntect")]
  let symbol_group_json_path = testdata_path.join("symbol_group-syntect.json");
  #[cfg(feature = "tree-sitter")]
  let symbol_group_json_path =
    testdata_path.join("symbol_group-tree-sitter.json");
  #[cfg(all(not(feature = "syntect"), not(feature = "tree-sitter")))]
  let symbol_group_json_path = testdata_path.join("symbol_group.json");

  if std::env::var("UPDATE").is_ok() {
    fs::write(&symbol_group_json_path, &files_json).unwrap();
  }

  let symbol_group_json = fs::read_to_string(symbol_group_json_path).unwrap();

  assert_eq!(files_json, symbol_group_json);
}

#[tokio::test]
async fn symbol_search() {
  let multiple_dir = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("multiple");

  let doc_nodes_by_url = get_files("multiple").await;

  let mut rewrite_map = IndexMap::new();
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
    ".".to_string(),
  );
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("b.ts")).unwrap(),
    "foo".to_string(),
  );

  let ctx = GenerateCtx::new(
    GenerateOptions {
      package_name: None,
      main_entrypoint: Some(
        ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
      ),
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: Some(rewrite_map),
      composable_output: false,
    },
    None,
    Default::default(),
    doc_nodes_by_url,
  )
  .unwrap();

  let search_index = generate_search_index(&ctx);
  let mut file_json = serde_json::to_string_pretty(&search_index).unwrap();
  file_json.push('\n');

  let symbol_search_json_path = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("symbol_search.json");

  if std::env::var("UPDATE").is_ok() {
    fs::write(&symbol_search_json_path, &file_json).unwrap();
  }

  let symbol_search_json = fs::read_to_string(symbol_search_json_path).unwrap();

  assert_eq!(file_json, symbol_search_json);
}

#[tokio::test]
async fn module_doc() {
  let multiple_dir = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata")
    .join("multiple");

  let doc_nodes_by_url = get_files("multiple").await;

  let mut rewrite_map = IndexMap::new();
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
    ".".to_string(),
  );
  rewrite_map.insert(
    ModuleSpecifier::from_file_path(multiple_dir.join("b.ts")).unwrap(),
    "foo".to_string(),
  );

  let ctx = GenerateCtx::new(
    GenerateOptions {
      package_name: None,
      main_entrypoint: Some(
        ModuleSpecifier::from_file_path(multiple_dir.join("a.ts")).unwrap(),
      ),
      href_resolver: Rc::new(EmptyResolver {}),
      usage_composer: None,
      rewrite_map: Some(rewrite_map),
      composable_output: false,
    },
    None,
    FileMode::Single,
    doc_nodes_by_url,
  )
  .unwrap();

  let mut module_docs = vec![];

  for (short_path, doc_nodes) in &ctx.doc_nodes {
    let render_ctx =
      RenderContext::new(&ctx, doc_nodes, UrlResolveKind::File(short_path));
    let module_doc = jsdoc::ModuleDocCtx::new(&render_ctx, short_path);

    module_docs.push(module_doc);
  }

  let mut file_json = serde_json::to_string_pretty(&module_docs).unwrap();
  file_json.push('\n');

  let testdata_path = std::env::current_dir()
    .unwrap()
    .join("tests")
    .join("testdata");

  #[cfg(feature = "syntect")]
  let module_docs_json_path = testdata_path.join("module_doc-syntect.json");
  #[cfg(feature = "tree-sitter")]
  let module_docs_json_path = testdata_path.join("module_doc-tree-sitter.json");
  #[cfg(all(not(feature = "syntect"), not(feature = "tree-sitter")))]
  let module_docs_json_path = testdata_path.join("module_doc.json");

  if std::env::var("UPDATE").is_ok() {
    fs::write(&module_docs_json_path, &file_json).unwrap();
  }

  let module_docs_json = fs::read_to_string(module_docs_json_path).unwrap();
  assert_eq!(file_json, module_docs_json);
}
