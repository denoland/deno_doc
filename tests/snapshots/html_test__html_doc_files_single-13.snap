---
source: tests/html_test.rs
expression: files.get(file_name).unwrap()
---
(function () {
  window.DENO_DOC_SEARCH_INDEX = {"kind":"search","nodes":[{"id":"namespace_Bar","kind":[{"char":"c","kind":"Class","title":"Class"}],"name":"Bar","file":".","doc":"","url":"././~/Bar.html","deprecated":false},{"id":"namespace_Foo","kind":[{"char":"c","kind":"Class","title":"Class"}],"name":"Foo","file":".","doc":"","url":"././~/Foo.html","deprecated":false},{"id":"namespace_Foobar","kind":[{"char":"c","kind":"Class","title":"Class"}],"name":"Foobar","file":".","doc":"","url":"././~/Foobar.html","deprecated":false}]};
})()
